# Data Science for Mobility / Intro to Business Analytics

## Lecture 2 - Numpy, statistics

Welcome back! :-) 

Today, we'll explore in more depth the possibilities providaded by the Python Numpy module. We lined up an incremental set of exercises that will end up with your first discriptive statistics analysis of real data!

**This notebook was designed to be done (or, at least, started) at the classroom. It's possible that mid-way or so, you'll feel comfortable to complete it by yourself, but we designed it such that we will be around most of the time. **

So, let's start. Since we are going to work with numpy, let's just import it, shall we?

In [1]:
data=open("pickups_zone_1_15min.csv",newline="\n").readlines()

We need some data. Please open the file "pickups_zone_1_15min.csv". This corresponds to the series of taxi-pickups in New York zone 1 (an area in the Manhattan island). 

You can use the function open(file), which **returns** a stream. 


In [2]:
len(data)

262849

Good, so we have the file stream, now we need to get to its content. This stream has a method to read all the lines at once (method readlines()). Let's use it. 

Just to be sure, let's check how many lines the file actually has (it should be 262849, if it's different, there's something wrong...). 

Print the first 10 lines of this file.

In [3]:
for i in data[:10]:
    print(i)

date,hour,minute,pickups

2009-01-01,0,0,0

2009-01-01,0,15,166

2009-01-01,0,30,215

2009-01-01,0,45,223

2009-01-01,1,0,245

2009-01-01,1,15,182

2009-01-01,1,30,181

2009-01-01,1,45,190

2009-01-01,2,0,197



Cool, so you have temporal attributes, and the actual number of pickps.

So, print **only** the pickups part. 

A tip - look what the following code does:

> x="1,2,3,4"

>xsplitted=x.split(',')


>print(xsplitted)

>print(xsplitted[2])

Output:

['1', '2', '3', '4']

3


In [4]:
data[0][:-1]

'date,hour,minute,pickups'

In [5]:
pickups=[]
for c,i in enumerate(data[1:]):
    i=data[c][:-1]
    data[c]=i.split(',')
    pickups.append(int(data[c][3]))

ValueError: invalid literal for int() with base 10: 'pickups'

In [ ]:
data[1:]

ok, our goal is thus to make a **single** list with all the pickup data. You need to go over each line, get the pickup value, convert it to an integer, and add it to this list. 

Can you do that?

Cool, so now you have an actual time series dataset. As you'll shortly see, it's very useful to have the respective time stamps (not just know that it is a sequence of values, also know exactly **when** each value occured)

Notice, however, that the values for the date are split into 3 fields. More importantly, for Python, they just numbers that have nothing to do with time. Luckily, there's class called datetime.datetime (the repetition here is intentional...). 

Here's some example code for you:
>from datetime import datetime as dt

>s="2017-09-11"
>
>print(type(s))
>
>time=dt.strptime(s,  '%Y-%m-%d')
>
>print(type(time))
>
>print(time)
>
>time=time.replace(hour=14, minute=35)
>
>print(time)


Output:


<class 'str' &gt;

<class 'datetime.datetime' &gt;

2017-09-11 00:00:00

2017-09-11 14:35:00





So, it creates a "datetime" object, that has everything you need to know about the time of that data point. It's pretty handy as you'll see later. 

Like what you did in the previous exercise, we now want a single list with all the datetime objects. 

Time to reorganize your code a bit. You've made a few things above:
- load a file
- go over the content of the file to create a list with pickup data
- go over the content of the file to create a list of datetime objects

Let's put them together in a function that reads a file (with name fscv) and returns the two lists mentioned. I'll give you the first and last lines:

>def read_csv(fcsv):
>
>      ...  #you just need to fill this part! ;-)
>
>      return pickups, times


Now, with this function, you can run all the above with different files with a single command! Do you want to try?


Now, let's create two numpy arrays from our two lists. 

The pickup vector can be used right away to make a histogram. Think about it: what should the distribution of number of pickups (i.e. the observable taxi demand) look like?

In [ ]:
import matplotlib.pyplot as plt   # we add this for you... 
#Yes, without this line below you'll run into trounble (wanna try?   ;-) )
%matplotlib inline       

hhmmm... was this what you were expecting?... 

Another interesting way to look at this data is simply by plotting directly with a scatter plot, where the x axis is the index of the datapoint, and the y axis is the total number of pickups.

Tip: to see better the data try playing with the size of the dots (for example, put s=0.1 as an argument to the scatter call)

This sort of pattern is intriguing. There seems to be two general "trends" of taxi pickups in this area. Do you think it relates with time of day? It would make some sense (e.g. during the night, low values, during peak hours, high values). 

There are many ways to check this, but a simple one is to make a small change in the temporal vector. Instead of each element of this vector corresponding to absolute time (the *actual* date and time), why not just represent the minutes since midnight? 

Can you make a new vector with that content?

**Tip on how to get the hour and minute**:
>from datetime import datetime as dt
>
>s="2017-09-11 18:35:11"
>
>d=dt.strptime(s, "%Y-%m-%d %H:%M:%S")
>
>print(d)
>
>print(d.hour)
>
>print(d.minute)
>

Output:

2017-09-11 18:35:11

18

35

Ok, now for a cool trick. In Python (well, in general), you ultimately define colors with numbers. So, imagine that the number of minutes since midnight (that you just created) corresponds to a color. The function scatter allows you to give this list straight away and plot it (just use the argument c, for example "c=my_minute_since_midnight_list". 

Do you want to try?

Doesn't this explain something?  ;-)

Ok, so there seems to be indeed a relationship with time! 

If this is true, it may be interesting to do a 24-hr average plot. In other words, a plot where the x axis is 0 to 1440 (1440=24 hours X 60 minutes), and you show the average per minute.

BTW, it will also be very useful if you add the 5 and 95 quantiles. 

Do you want to do this?

Ok, again, you did a lot of stuff. Why not reorganize it again?

For example, a new function that receives the file name, and generates all the graphs you did so far.

Wouldn't it be cool to just generate those by a single command?

### Correlations

One very important task in Data Science modeling is to find (and understand) correlations between different variables. Let's do a few simple exercises.

Let's start with a simple question: are the different areas correlated between them? If yes, it may be interesting knowledge. For example, maybe we can share data between them to predict better, later.



In [ ]:
s1,_ =read_csv("pickups_zone_1_15min.csv")
s17, _=read_csv("pickups_zone_17_15min.csv")
s21, _=read_csv("pickups_zone_21_15min.csv")
s28, _=read_csv("pickups_zone_28_15min.csv")

np.corrcoef([s1, s17, s21, s28])

It seems all but area 28 are well correlated with each other. Can that show, in the pictures above?

Now, a more interesting question: are there correlations between a given area, and the other areas in earlier time steps? 

This is a VERY important one. If you find high correlation, for example, between area 1 at time t, with area 17 at time t-1, then you can use area 17 to predict area 1!

To check this, you need to play a little bit with the vectors. Let's call a vector that is shifted in time for 1 time step, a "lag1" vector.:

In [ ]:
s1_trimmed=np.array(s1[1:]) #look, it's the SAME vector, except that you start on the second element. Why?
s21_lag1=np.array(s21[:-1]) #if this is confusing you now, please ask the teacher to explain. 
s17_lag1=np.array(s17[:-1])#With a white board near, it will be easy to understand! ;-)

Let's now check those correlations

In [ ]:
np.corrcoef([s1_trimmed, s17_lag1, s21_lag1])

WOW! Very interesting!! This means that you can use data from these other areas to predict for area 1... This is useful when there is missing data in area 1, for example... 

Try to replicate this same study, for all different areas. To make it quick, you can define a function that receives the area you want to "predict for", and the areas you want to use as "predictors".

An idea: check the correlation between the series "s1" and the series "s1_lag1". This is called the autocorrelation of lag 1 of the time series. 

Now, do this with lag=2

Ok, the last couple of tasks for today: 

Create a list, let's call it "auto_s1", with the autocorrelations with all lags until 120

Now, make a plot such that The x axis is the range(1,120), and y axis is the auto_s1 list. 

The resulting diagram is called an autocorrelogram! :-) What do you think? Can you explain what's happening in those bumps?